### PR TITLE
[SMTChecker] Resolve current contract context correctly in VariableUsage

### LINKED ISSUE #10926: [SMTChecker] ICE in auto solidity::frontend::SMTEncoder::functionCallToDefinition(const solidity::frontend::FunctionCall &, const solidity::frontend::ContractDefinition *, const solidity::frontend::ContractDefinition *)::(anonymous class)::operator()(const auto *) const [_ref:auto = solidity::frontend::MemberAccess]
## Description

This contract:
```solidity
pragma experimental SMTChecker;
contract A {
    function f() internal virtual {
	v();
    }
    function v() internal virtual {
    }
}

contract B is A {
    function f() internal virtual override {
	super.f();
    }
}

contract C is B {
    function v() internal override {
	if (0==1)
	    f();
    }
}
```

produces:

```
Internal compiler error during compilation:
/solidity/libsolidity/formal/SMTEncoder.cpp(2722): Throw in function auto solidity::frontend::SMTEncoder::functionCallToDefinition(const solidity::frontend::FunctionCall &, const solidity::frontend::ContractDefinition *, const solidity::frontend::ContractDefinition *)::(anonymous class)::operator()(const auto *) const [_ref:auto = solidity::frontend::MemberAccess]
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: Super contract not available.
[solidity::util::tag_comment*] = Super contract not available.
```

when compiled with solc.

Looks like all other deaths in isDynamicallyEncoded were fixed, so this is a new variant (the structure looks different to me at a glance).

On master, using AFL fuzzing.  Another discovery using https://github.com/agroce/afl-compiler-fuzzer.

## Environment

- Compiler version:  0.8.2-develop.2021.2.9+commit.43319574.Linux.clang
- Target EVM version (as per compiler settings): N/A
- Framework/IDE (e.g. Truffle or Remix): N/A
- EVM execution environment / backend / blockchain client: N/A
- Operating system: Ubuntu 20.04.1 LTS in docker

## Steps to Reproduce

Above shows pretty clearly, I think.

### LINKED ISSUE #10957: [SMTChecker] ICE in FunctionDefinition::resolveVirtual(): "Virtual function _approve not found"
From https://github.com/ethereum/solidity/pull/10952#issuecomment-778307247.

## Description
This contract triggers an ICE in `soltest` when placed under `tests/smtCheckerTests/`:

``` solidity
pragma experimental SMTChecker;

contract Context {}

contract ERC20 is Context {
    function approve() public virtual { _approve(); }
    function _approve() internal virtual {}
}

contract __unstable__ERC20Owned is ERC20 {
    function _approve() internal override {
        if (true) {
            super._approve();
        }
    }
}
```

Output:
```
/solidity/test/boostTest.cpp(123): error: in "smtCheckerTests/inheritance/gsn": Exception during extracted test: /solidity/libsolidity/ast/AST.cpp(394): Throw in function virtual const solidity::frontend::FunctionDefinition& solidity::frontend::FunctionDefinition::resolveVirtual(const solidity::frontend::ContractDefinition&, const solidity::frontend::ContractDefinition*) const
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: Virtual function _approve not found.
[solidity::util::tag_comment*] = Virtual function _approve not found.
```

The code comes originally from OpenZeppelin. `__unstable__ERC20Owned` is defined in `contracts/GSN/GSNRecipientERC20Fee.sol` in OpenZeppelin. In the above I have removed everything that wasn't relevant and combined it with `ERC20` and `Context`.

## Environment
- Compiler version: `0.8.2-develop.2021.2.3+commit.2fa2b8b7.mod.Linux.g++`

### PR DESCRIPTION
**Pull request by blishko**
Fixes #10957.
Fixes #10926.

### CODE REVIEW

File: libsolidity/formal/SMTEncoder.cpp [hunk 1]
```diff
@@ -177,8 +177,12 @@ void SMTEncoder::visitFunctionOrModifier()
 	{
 		if (function.isImplemented())
 		{
+			auto const* currentFunction = m_variableUsage.getCurrentFunction();
+			solAssert(currentFunction, "");
+			m_variableUsage.setCurrentFunction(function);
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]: So BMC's function inlining ends up here?

Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: Yes, exactly.
The problem is that `VariableUsage` previously did not update what is the current function and then used incorrect `scopeContract` in `VariableUsage::endVisit(FunctionCall const&` leading to a lookup of a function in a contract that did not have it.

File: libsolidity/formal/VariableUsage.h [hunk 2]
```diff
@@ -41,6 +41,8 @@ class VariableUsage: private ASTConstVisitor
 	void setCurrentContract(ContractDefinition const& _contract) { m_currentContract = &_contract; }
 	void setCurrentFunction(FunctionDefinition const& _function) { m_currentFunction = &_function; }
 
+	FunctionDefinition const* getCurrentFunction() { return m_currentFunction; }
```
Comment (leonardoalt) [inline libsolidity/formal/VariableUsage.h]:
```suggestion
	FunctionDefinition const* currentFunction() { return m_currentFunction; }
```

**Comment by blishko** [comment]
@leonardoalt , I think I have a more elegant fix for #10957 than the previous one.
Can you check it if you agree? I just changed how current scope contract is detected, taking the idea from `SMTEncoder::currentScopeContract`. If yes, I will squash this and add a changelog entry.

**Comment by leonardoalt** [comment]
Ah nice, looks good indeed!

**Comment by blishko** [comment]
We could even remove the field `m_currentFunction` from `VariableUsage`. After this change, it will not be used anymore. Should we remove it or do you think it might still be useful? I would remove it, since we know that in some situations it might not hold the right value.

**Comment by leonardoalt** [comment]
Yea true, I was thinking the same. Please remove it if possible.

File: libsolidity/formal/VariableUsage.h (hunk 2 above)
Comment (blishko) [reply libsolidity/formal/VariableUsage.h]: This method is not needed anymore. The issue is resolved in a different way.

**Review by leonardoalt — DISMISSED** [review]

**Comment by blishko** [comment]
I have just tested also #10926 and this PR fixes also that issue. I will add it to the regression tests.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**